In [53]:
import pandas as pd
#gender-classifier-DFE-791531.csv


In [54]:
df = pd.read_csv("C:\\Users\In9.var\Desktop\Projects\data\gender-classifier-DFE-791531.csv", engine='python')
df

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0      815719226    False   finalized                   3    10/26/15 23:24   
1      815719227    False   finalized                   3    10/26/15 23:30   
2      815719228    False   finalized                   3    10/26/15 23:33   
3      815719229    False   finalized                   3    10/26/15 23:10   
4      815719230    False   finalized                   3     10/27/15 1:15   
...          ...      ...         ...                 ...               ...   
20045  815757572     True      golden                 259               NaN   
20046  815757681     True      golden                 248               NaN   
20047  815757830     True      golden                 264               NaN   
20048  815757921     True      golden                 250               NaN   
20049  815757985     True      golden                 249               NaN   

       gender  gender:confidence profile_yn  profile_yn:confidence  \
0        male             1.0000        yes                    1.0   
1        male             1.0000        yes                    1.0   
2        male             0.6625        yes                    1.0   
3        male             1.0000        yes                    1.0   
4      female             1.0000        yes                    1.0   
...       ...                ...        ...                    ...   
20045  female             1.0000        yes                    1.0   
20046    male             1.0000        yes                    1.0   
20047    male             1.0000        yes                    1.0   
20048  female             0.8489        yes                    1.0   
20049  female             1.0000        yes                    1.0   

              created  ...                                       profileimage  \
0        12/5/13 1:48  ...  https://pbs.twimg.com/profile_images/414342229...   
1       10/1/12 13:51  ...  https://pbs.twimg.com/profile_images/539604221...   
2      11/28/14 11:30  ...  https://pbs.twimg.com/profile_images/657330418...   
3       6/11/09 22:39  ...  https://pbs.twimg.com/profile_images/259703936...   
4       4/16/14 13:23  ...  https://pbs.twimg.com/profile_images/564094871...   
...               ...  ...                                                ...   
20045    8/5/15 21:16  ...  https://pbs.twimg.com/profile_images/656793310...   
20046   8/15/12 21:17  ...  https://pbs.twimg.com/profile_images/639815429...   
20047     9/3/12 1:17  ...  https://pbs.twimg.com/profile_images/655473271...   
20048   11/6/12 23:46  ...  https://pbs.twimg.com/profile_images/657716093...   
20049   4/14/14 17:22  ...  https://pbs.twimg.com/profile_images/655134724...   

       retweet_count sidebar_color  \
0                  0        FFFFFF   
1                  0        C0DEED   
2                  1        C0DEED   
3                  0        C0DEED   
4                  0             0   
...              ...           ...   
20045              0        C0DEED   
20046              0             0   
20047              0        C0DEED   
20048              0             0   
20049              0        C0DEED   

                                                    text tweet_coord  \
0      Robbie E Responds To Critics After Win Against...         NaN   
1      ‰ЫПIt felt like they were my friends and I was...         NaN   
2      i absolutely adore when louis starts the songs...         NaN   
3      Hi @JordanSpieth - Looking at the url - do you...         NaN   
4      Watching Neighbours on Sky+ catching up with t...         NaN   
...                                                  ...         ...   
20045  @lookupondeath ...Fine, and I'll drink tea too...         NaN   
20046  Greg Hardy you a good player and all but don't...         NaN   
20047  You can miss people and still never want to se...         NaN   
20048  @bitemyapp i had noticed your tendency to pee ...         NaN

In [55]:
df.isna().sum()

_unit_id                     0
_golden                      0
_unit_state                  0
_trusted_judgments           0
_last_judgment_at           50
gender                      97
gender:confidence           26
profile_yn                   0
profile_yn:confidence        0
created                      0
description               3744
fav_number                   0
gender_gold              20000
link_color                   0
name                         0
profile_yn_gold          20000
profileimage                 0
retweet_count                0
sidebar_color                0
text                         0
tweet_coord              19891
tweet_count                  0
tweet_created                0
tweet_id                     0
tweet_location            7484
user_timezone             7798
dtype: int64

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20050 entries, 0 to 20049
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   _unit_id               20050 non-null  int64  
 1   _golden                20050 non-null  bool   
 2   _unit_state            20050 non-null  object 
 3   _trusted_judgments     20050 non-null  int64  
 4   _last_judgment_at      20000 non-null  object 
 5   gender                 19953 non-null  object 
 6   gender:confidence      20024 non-null  float64
 7   profile_yn             20050 non-null  object 
 8   profile_yn:confidence  20050 non-null  float64
 9   created                20050 non-null  object 
 10  description            16306 non-null  object 
 11  fav_number             20050 non-null  int64  
 12  gender_gold            50 non-null     object 
 13  link_color             20050 non-null  object 
 14  name                   20050 non-null  object 
 15  pr

In [57]:
df = df.drop(['tweet_coord', 'gender_gold', 'profile_yn_gold', '_last_judgment_at', 'profileimage', 'description', 'text', 'user_timezone', 'tweet_location', '_unit_id', 'created', 'tweet_id', 'tweet_created', 'name', 'link_color', 'sidebar_color'], axis=1) # абсолюто бесполезные данные

In [58]:
df.info()
df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20050 entries, 0 to 20049
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   _golden                20050 non-null  bool   
 1   _unit_state            20050 non-null  object 
 2   _trusted_judgments     20050 non-null  int64  
 3   gender                 19953 non-null  object 
 4   gender:confidence      20024 non-null  float64
 5   profile_yn             20050 non-null  object 
 6   profile_yn:confidence  20050 non-null  float64
 7   fav_number             20050 non-null  int64  
 8   retweet_count          20050 non-null  int64  
 9   tweet_count            20050 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(3)
memory usage: 1.4+ MB


_golden _unit_state  _trusted_judgments  gender  gender:confidence  \
0        False   finalized                   3    male             1.0000   
1        False   finalized                   3    male             1.0000   
2        False   finalized                   3    male             0.6625   
3        False   finalized                   3    male             1.0000   
4        False   finalized                   3  female             1.0000   
...        ...         ...                 ...     ...                ...   
20045     True      golden                 259  female             1.0000   
20046     True      golden                 248    male             1.0000   
20047     True      golden                 264    male             1.0000   
20048     True      golden                 250  female             0.8489   
20049     True      golden                 249  female             1.0000   

      profile_yn  profile_yn:confidence  fav_number  retweet_count  \
0            yes                    1.0           0              0   
1            yes                    1.0          68              0   
2            yes                    1.0        7696              1   
3            yes                    1.0         202              0   
4            yes                    1.0       37318              0   
...          ...                    ...         ...            ...   
20045        yes                    1.0         302              0   
20046        yes                    1.0        2513              0   
20047        yes                    1.0          50              0   
20048        yes                    1.0       70804              0   
20049        yes                    1.0        3348              0   

       tweet_count  
0           110964  
1             7471  
2             5617  
3             1693  
4            31462  
...            ...  
20045          783  
20046        13523  
20047        26419  
20048        56073  
20049         2922  

[20050 rows x 10 columns]

In [59]:
df.isna().sum()

_golden                   0
_unit_state               0
_trusted_judgments        0
gender                   97
gender:confidence        26
profile_yn                0
profile_yn:confidence     0
fav_number                0
retweet_count             0
tweet_count               0
dtype: int64

In [60]:
df = df.dropna()

In [61]:
df = pd.get_dummies(df, columns=['_unit_state', 'profile_yn'])
df

_golden  _trusted_judgments  gender  gender:confidence  \
0        False                   3    male             1.0000   
1        False                   3    male             1.0000   
2        False                   3    male             0.6625   
3        False                   3    male             1.0000   
4        False                   3  female             1.0000   
...        ...                 ...     ...                ...   
20045     True                 259  female             1.0000   
20046     True                 248    male             1.0000   
20047     True                 264    male             1.0000   
20048     True                 250  female             0.8489   
20049     True                 249  female             1.0000   

       profile_yn:confidence  fav_number  retweet_count  tweet_count  \
0                        1.0           0              0       110964   
1                        1.0          68              0         7471   
2                        1.0        7696              1         5617   
3                        1.0         202              0         1693   
4                        1.0       37318              0        31462   
...                      ...         ...            ...          ...   
20045                    1.0         302              0          783   
20046                    1.0        2513              0        13523   
20047                    1.0          50              0        26419   
20048                    1.0       70804              0        56073   
20049                    1.0        3348              0         2922   

       _unit_state_finalized  _unit_state_golden  profile_yn_yes  
0                          1                   0               1  
1                          1                   0               1  
2                          1                   0               1  
3                          1                   0               1  
4                          1                   0               1  
...                      ...                 ...             ...  
20045                      0                   1               1  
20046                      0                   1               1  
20047                      0                   1               1  
20048                      0                   1               1  
20049                      0                   1               1  

[19953 rows x 11 columns]

In [62]:
df[df['gender'] == 'brand'] = None
df[df['gender'] == 'unknown'] = None
df = df.dropna()
df

_golden  _trusted_judgments  gender  gender:confidence  \
0          0.0                 3.0    male             1.0000   
1          0.0                 3.0    male             1.0000   
2          0.0                 3.0    male             0.6625   
3          0.0                 3.0    male             1.0000   
4          0.0                 3.0  female             1.0000   
...        ...                 ...     ...                ...   
20045      1.0               259.0  female             1.0000   
20046      1.0               248.0    male             1.0000   
20047      1.0               264.0    male             1.0000   
20048      1.0               250.0  female             0.8489   
20049      1.0               249.0  female             1.0000   

       profile_yn:confidence  fav_number  retweet_count  tweet_count  \
0                        1.0         0.0            0.0     110964.0   
1                        1.0        68.0            0.0       7471.0   
2                        1.0      7696.0            1.0       5617.0   
3                        1.0       202.0            0.0       1693.0   
4                        1.0     37318.0            0.0      31462.0   
...                      ...         ...            ...          ...   
20045                    1.0       302.0            0.0        783.0   
20046                    1.0      2513.0            0.0      13523.0   
20047                    1.0        50.0            0.0      26419.0   
20048                    1.0     70804.0            0.0      56073.0   
20049                    1.0      3348.0            0.0       2922.0   

       _unit_state_finalized  _unit_state_golden  profile_yn_yes  
0                        1.0                 0.0             1.0  
1                        1.0                 0.0             1.0  
2                        1.0                 0.0             1.0  
3                        1.0                 0.0             1.0  
4                        1.0                 0.0             1.0  
...                      ...                 ...             ...  
20045                    0.0                 1.0             1.0  
20046                    0.0                 1.0             1.0  
20047                    0.0                 1.0             1.0  
20048                    0.0                 1.0             1.0  
20049                    0.0                 1.0             1.0  

[12894 rows x 11 columns]

####нужно ещё преобразовать male female в 0 1

In [63]:
# Используем LabelEncoder, тк это будет целевой переменной
from sklearn import preprocessing as pre

l = pre.LabelEncoder()
mapped = l.fit_transform(df['gender'])
df['gender'] = mapped
df

<ipython-input-63-7ab908fd9105>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = mapped


_golden  _trusted_judgments  gender  gender:confidence  \
0          0.0                 3.0       1             1.0000   
1          0.0                 3.0       1             1.0000   
2          0.0                 3.0       1             0.6625   
3          0.0                 3.0       1             1.0000   
4          0.0                 3.0       0             1.0000   
...        ...                 ...     ...                ...   
20045      1.0               259.0       0             1.0000   
20046      1.0               248.0       1             1.0000   
20047      1.0               264.0       1             1.0000   
20048      1.0               250.0       0             0.8489   
20049      1.0               249.0       0             1.0000   

       profile_yn:confidence  fav_number  retweet_count  tweet_count  \
0                        1.0         0.0            0.0     110964.0   
1                        1.0        68.0            0.0       7471.0   
2                        1.0      7696.0            1.0       5617.0   
3                        1.0       202.0            0.0       1693.0   
4                        1.0     37318.0            0.0      31462.0   
...                      ...         ...            ...          ...   
20045                    1.0       302.0            0.0        783.0   
20046                    1.0      2513.0            0.0      13523.0   
20047                    1.0        50.0            0.0      26419.0   
20048                    1.0     70804.0            0.0      56073.0   
20049                    1.0      3348.0            0.0       2922.0   

       _unit_state_finalized  _unit_state_golden  profile_yn_yes  
0                        1.0                 0.0             1.0  
1                        1.0                 0.0             1.0  
2                        1.0                 0.0             1.0  
3                        1.0                 0.0             1.0  
4                        1.0                 0.0             1.0  
...                      ...                 ...             ...  
20045                    0.0                 1.0             1.0  
20046                    0.0                 1.0             1.0  
20047                    0.0                 1.0             1.0  
20048                    0.0                 1.0             1.0  
20049                    0.0                 1.0             1.0  

[12894 rows x 11 columns]

In [64]:
#Делаем масштабирование признаков: нормализацию и стандартизацию
X = df.drop(['gender'], axis = 1)
y = df['gender']
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_st = scaler.fit_transform(X)
print(X_st)
print('###################'*5)
scaler = MinMaxScaler()
X_norm = scaler.fit_transform(X)

[[ -0.05291326  -0.05285441   0.50697159 ...   0.05291326  -0.05291326
    0.        ]
 [ -0.05291326  -0.05285441   0.50697159 ...   0.05291326  -0.05291326
    0.        ]
 [ -0.05291326  -0.05285441  -1.53369911 ...   0.05291326  -0.05291326
    0.        ]
 ...
 [ 18.89885358  19.92162621   0.50697159 ... -18.89885358  18.89885358
    0.        ]
 [ 18.89885358  18.85019813  -0.40664424 ... -18.89885358  18.89885358
    0.        ]
 [ 18.89885358  18.77366756   0.50697159 ... -18.89885358  18.89885358
    0.        ]]
###############################################################################################


In [32]:
#svm с обычными данными + кросвалидация + оценка
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.svm import SVC
model=SVC()
cvs0 = cross_val_score(estimator=model, X=X, y=y, cv=6, scoring='accuracy')
cvs1 = cross_val_score(estimator=model, X=X, y=y, cv=6, scoring='recall')
cvs2 = cross_val_score(estimator=model, X=X, y=y, cv=6, scoring='precision')
print("accuracy")
print(cvs0.mean())
print("recall")
print(cvs1.mean())
print("precision")
print(cvs2.mean())

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision 

accuracy
0.5248177446874515
recall
0.042312661498708014
precision
0.09555069292487235


In [33]:
#svm с нормализированными данными + кросвалидация + оценка
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.svm import SVC
model=SVC()
cvs0 = cross_val_score(estimator=model, X=X_norm, y=y, cv=6, scoring='accuracy')
cvs1 = cross_val_score(estimator=model, X=X_norm, y=y, cv=6, scoring='recall')
cvs2 = cross_val_score(estimator=model, X=X_norm, y=y, cv=6, scoring='precision')
print("accuracy")
print(cvs0.mean())
print("recall")
print(cvs1.mean())
print("precision")
print(cvs2.mean())

accuracy
0.534977508918877
recall
0.23620319507918283
precision
0.5356822544699683


In [34]:
#svm с стандартизированными данными + кросвалидация + оценка
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.svm import SVC
model=SVC()
cvs0 = cross_val_score(estimator=model, X=X_st, y=y, cv=6, scoring='accuracy')
cvs1 = cross_val_score(estimator=model, X=X_st, y=y, cv=6, scoring='recall')
cvs2 = cross_val_score(estimator=model, X=X_st, y=y, cv=6, scoring='precision')
print("accuracy")
print(cvs0.mean())
print("recall")
print(cvs1.mean())
print("precision")
print(cvs2.mean())

accuracy
0.5361408407011014
recall
0.24169415115153425
precision
0.5369997364121583


In [65]:
#knn с обычными данными
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=1/3, random_state=42)
knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
confusion_matrix(y_test,y_pred)

print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

Accuracy: 0.5351326198231736
Recall: 0.515972894482091
Precision: 0.5164728682170543


In [66]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(knn, X, y, cv=10)
print(cvs)
print(cvs.mean())

[0.51395349 0.52325581 0.51472868 0.52015504 0.55702095 0.53529868
 0.5477114  0.51280062 0.51202483 0.5321955 ]
0.526914500153355


In [67]:
#knn со стандартизированными данными
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)
knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
confusion_matrix(y_test,y_pred)

print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

Accuracy: 0.5379246161005119
Recall: 0.489351403678606
Precision: 0.5205973223480947


In [68]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(knn, X_st, y, cv=10)
print(cvs)
print(cvs.mean())

[0.50852713 0.52713178 0.52868217 0.52093023 0.54460822 0.50814585
 0.53374709 0.52211016 0.51512801 0.52986811]
0.5238878765463283


In [69]:
#knn с нормализованными данными
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
confusion_matrix(y_test,y_pred)

print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

Accuracy: 0.5435086086551885
Recall: 0.4912875121006776
Precision: 0.5269989615784009


In [70]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(knn, X_norm, y, cv=10)
print(cvs)
print(cvs.mean())

[0.52170543 0.54031008 0.53565891 0.53410853 0.54538402 0.51512801
 0.50969744 0.54072925 0.52211016 0.51745539]
0.5282287212609981


In [71]:
#Логистическая регрессия с обычными данными
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=1/3)

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(model.score(X_test, y_test))
print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

0.547463936714751
Accuracy: 0.547463936714751
Recall: 0.42974806201550386
Precision: 0.5359516616314199


In [72]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(model, X, y, cv=10)
print(cvs)
print(cvs.mean())

[0.54108527 0.53953488 0.53565891 0.55891473 0.57020946 0.55624515
 0.55702095 0.54615981 0.54926299 0.52443755]
0.5478529717766912


In [73]:
#Логистическая регрессия со стандартизированными данными
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_st,y, test_size=1/3)

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(model.score(X_test, y_test))
print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

0.5381572824569567
Accuracy: 0.5381572824569567
Recall: 0.2503617945007236
Precision: 0.5463157894736842


In [74]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(model, X_st, y, cv=10)
print(cvs)
print(cvs.mean())

[0.53565891 0.53023256 0.51627907 0.52248062 0.55624515 0.53762607
 0.54305663 0.54926299 0.55546936 0.56322731]
0.5409538672488138


In [75]:
#Логистическая регрессия с нормализованными данными
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm,y, test_size=1/3)

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(model.score(X_test, y_test))
print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

0.5486272684969753
Accuracy: 0.5486272684969753
Recall: 0.2540381791483113
Precision: 0.5550802139037433


In [76]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(model, X_norm, y, cv=10)
print(cvs)
print(cvs.mean())

[0.53565891 0.52868217 0.51782946 0.51782946 0.55236618 0.54228084
 0.54072925 0.54693561 0.54926299 0.55469356]
0.5386268425135764


In [77]:
##Дерево решений с обычными
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=4, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.5691019078641228
Recall: 0.5130687318489835
Precision: 0.5561385099685204


In [48]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(tree, X, y, cv=10)
print(cvs)
print(cvs.mean())

[0.52790698 0.52713178 0.52170543 0.51317829 0.5484872  0.54072925
 0.52831652 0.54305663 0.54460822 0.55624515]
0.5351365459673685


In [49]:
#Дерево решений с нормализованными
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=4, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.539320614239181
Recall: 0.24878993223620524
Precision: 0.5456475583864119


In [50]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(tree, X_norm, y, cv=10)
print(cvs)
print(cvs.mean())

[0.52790698 0.52713178 0.52170543 0.51317829 0.5484872  0.54072925
 0.52831652 0.54305663 0.54460822 0.55624515]
0.5351365459673685


In [51]:
#Дерево решений со стандартизированными
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=4, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.539320614239181
Recall: 0.24878993223620524
Precision: 0.5456475583864119


In [52]:
#крос валидация
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(tree, X_st, y, cv=10)
print(cvs)
print(cvs.mean())

[0.52790698 0.52713178 0.52170543 0.51317829 0.5484872  0.54072925
 0.52831652 0.54305663 0.54460822 0.55624515]
0.5351365459673685


Выводы: 
1)наилучшим образом показало себя дерево решений Accuracy: 0.5691019078641228
2)метод опорных векторов нереально медленный
3)метод опорных векторов ситуативный и это не та ситуация (и плохие данные): recall 0.05424417979293145
4)Precision дерева решений не зависит от маштабирования(на этих данных) = 0.5561385099685204